nomad/structs/diff.go

```diff
@@ -571,9 +573,152 @@ func (t *Task) Diff(other *Task, contextual bool) (*TaskDiff, error) {
 		diff.Objects = append(diff.Objects, altIDDiffs...)
 	}
 
+	// Actions diff
+	if aDiffs := actionDiffs(t.Actions, other.Actions, contextual); aDiffs != nil {
+		diff.Objects = append(diff.Objects, aDiffs...)
+	}
+
 	return diff, nil
 }
 
+func findActionMatch(action *Action, newActions []*Action, newActionMatches []int) int {
+	indexMatch := -1
+
+	for i, newAction := range newActions {
+		if newActionMatches[i] >= 0 {
+			continue
+		}
+
+		if action.Name == newAction.Name {
+			indexMatch = i
+			break
+		}
+	}
+
+	return indexMatch
+}
+
+func actionDiff(old, new *Action, contextual bool) *ObjectDiff {
+	// Initialize a new ObjectDiff for this pair of actions
+	diff := &ObjectDiff{
+		Type:   "Action",
+		Name:   "", // populate based on your field
+		Fields: make([]*FieldDiff, 0),
+	}
+
+	// Assigning safe default values to compare
+	oldName := ""
+	newName := ""
+	oldCommand := ""
+	newCommand := ""
+	var oldArgsJSON, newArgsJSON []byte
+
+	// Populate values if old is not nil
+	if old != nil {
+		oldName = old.Name
+		oldCommand = old.Command
+		oldArgsJSON, _ = json.Marshal(old.Args)
+	}
+
+	// Populate values if new is not nil
+	if new != nil {
+		newName = new.Name
+		newCommand = new.Command
+		newArgsJSON, _ = json.Marshal(new.Args)
+	}
+
+	// Detect the type of difference: Added, Edited, Deleted
+	if old == nil && new != nil {
+		diff.Type = DiffTypeAdded
+		diff.Name = newName
+	} else if old != nil && new == nil {
+		diff.Type = DiffTypeDeleted
+		diff.Name = oldName
+	} else {
+		diff.Type = DiffTypeEdited
+		diff.Name = oldName // assuming names are constant
+	}
+
+	// Compare Command field
+	if oldCommand != newCommand || contextual {
+		field := &FieldDiff{
+			Type: "string",
+			Name: "Command",
+			Old:  oldCommand,
+			New:  newCommand,
+		}
+		diff.Fields = append(diff.Fields, field)
+	}
+
+	// Compare Args field
+	if !bytes.Equal(oldArgsJSON, newArgsJSON) || contextual {
+		field := &FieldDiff{
+			Type: "[]string",
+			Name: "Args",
+			Old:  string(oldArgsJSON),
+			New:  string(newArgsJSON),
+		}
+		diff.Fields = append(diff.Fields, field)
+	}
```

Comment:
Also in the `changeScriptDiff` function you can see the use of the `stringSetDiff` helper, which is perfect for comparing a slice of strings for `args`. The name is a bit of a misnomer, it's really a `stringSliceDiff` because it cares about order, which is what you want here.